### ISSUE TITLE
Argument slots not perserved with slots = true? 

### ISSUE THREAD

**Issue by phipsgabler** [MikeInnes/IRTools.jl#31]
As we see in the following example, the internal variable `y` is represented as a slot, but the argument `x` is just converted directly to `%2`:

```
julia> f3(x) = (y = x + 1; sin(2y))
f3 (generic function with 1 method)

julia> IRTools.IR(typeof(f3), Int; slots = true)
1: (%1, %2)
  %3 = %2 + 1
  %4 = @_3 = %3
  %5 = 2 * @_3
  %6 = Main.sin(%5)
  return %6


julia> @code_lowered f3(1)
CodeInfo(
1 ─      y = x + 1
│   %2 = 2 * y
│   %3 = Main.sin(%2)
└──      return %3
)
```

That appears to be done [on purpose](https://github.com/MikeInnes/IRTools.jl/blob/cfda7ffa9683e6ccf45589db83d160b35ffe8461/src/ir/wrap.jl#L171), but why?  Is it just because I can "trivially" get the argument slot names from `meta.code`?

**Comment by MikeInnes** [comment]
The first `n` slots in the CodeInfo are not real slots, in the sense of mutable locals; if you actually modify `x` you'll see that there's an argument slot (maps to our basic block arguments) as well as a slot for the local variable `x`. Conversely, with no mutation, there's no mutable local `x`.

```julia
julia> f(x) = (x += 1)
f (generic function with 1 method)

julia> @code_lowered f(1)
CodeInfo(
1 ─      x@_3 = x@_2
│   %2 = x@_3 + 1
│        x@_3 = %2
└──      return %2
)
```

The alternatives I can think of are to (1) follow codeinfo and remove explicit arguments in favour of a set of slots with default values, or (2) to have a slot for each argument that gets explicitly updated (`@x = %2`) in the first block. The former defeats SSA to some extent though, and the latter is much noisy without a lot of obvious benefit.

**Comment by phipsgabler** [comment]
Hm, OK.  I'm only using the slotnames as metadata, anyway.  I would have hoped to get something like this:

```
julia> IRTools.IR(typeof(f3), Int; slots = true)
1: (%1(#self#), %2(x))
  %3(y) = %2(x) + 1
  %5 = 2 * %3(y)
  %6 = Main.sin(%5)
  return %6
```

I guess this information is available at some point in the pass you're using to remove slots -- maybe [here](https://github.com/MikeInnes/IRTools.jl/blob/master/src/passes/passes.jl#L175)?

**Comment by MikeInnes** [comment]
Yeah, we at some point need to preserve block arguments names for things like debug information. I don't think this is too difficult to do, though it's a bit fiddly and potentially touches a bunch of code.
